### PR TITLE
Rework Sotw Cache interface

### PR DESCRIPTION
**Pull request by valerian-roche**
Datadog fork PR matching https://github.com/envoyproxy/go-control-plane/pull/584

It has been raised that the addition of StreamState within the Cache interface is breaking the separation of concern between the server side managing the xds protocol itself and the cache which is only in charge of returning responses based on watches
PR https://github.com/envoyproxy/go-control-plane/pull/577 has been opened proposing to fully remove any client state from the CreateWatch interface

This PR is an alternate proposal based on a few concerns/aspects:
- one of the argument on removing it is that the sotw protocol is stateless. It actually is not, and is explicitly visible in this PR: the lastResponse latched value as well as the version hack when the resource list is changed very well show that it is stateful
- in cases outside of lds and cds, it is clearly stated that the control-plane is not expected to send back all resources when a single one change. This is highly critical in the case of ads with eds, as there is a likelihood of both having a lot of resources and a high churn. The implementation sending everything each time is technically tolerated, but is pushing on the data-plane work that could be done at low cost on the control-plane
- the current support of both delta and sotw, for both linear and simple cache means that every change has to be done 4 times, with four very different implementations, each one supporting only a subset of use cases (e.g. linear does not support cds/lds as it doesn't handle wildcard in this case, and sotw does not work properly as resource list is improperly maintained). I strongly believed that both implementation should converge. The only remaining differences should be in the messages themselves, abstracted in the server part. Having a common interface for sotw and delta is a first step for this. It would also allow in the future to make snapshot far more efficient by supporting a linear version per snapshot

In this context, this PR is changing:
- creates a ClientState within the cache package, driven by the data needed from the cache perspective to be able to answer a client request
- simplifies streamState
  - makes it compatible with the new ClientState interface
  - removes the IsFirst notion, as the protocol states it should come from the nonce in the request
  - removes the duplicate/parallel knownResourceNames and use the delta one instead. It will also properly carry the version in a future PR

### CODE REVIEW

File: pkg/cache/v3/cache.go
Comment (atollena) [inline pkg/cache/v3/cache.go]: Small nit: While the term "stream state" made sense to disambiguate with just "stream", here I think you could get away with just `type Subscription interface`.

File: pkg/cache/v3/cache.go
Comment (atollena) [inline pkg/cache/v3/cache.go]:
```suggestion
	AckedResources() map[string]string
```

See https://go.dev/doc/effective_go#Getters.

File: pkg/cache/v3/cache.go
Comment (atollena) [inline pkg/cache/v3/cache.go]:
```suggestion
	SubscribedResources() map[string]struct{}
```

File: pkg/cache/v3/cache.go
Comment (atollena) [inline pkg/cache/v3/cache.go:58]: Same, could just be `Wildcard() bool`

File: pkg/cache/v3/linear.go
Comment (atollena) [inline pkg/cache/v3/linear.go:303]: Are there practical cases where this is important to catch this in a production system?

My understanding is that this would be the result of a muxer that is not configured properly, IMO it's not worth increasing the size of the interface for all the valid cases. I'll leave it up to you though.

File: pkg/cache/v3/mux.go
Comment (atollena) [inline pkg/cache/v3/mux.go:43]: This is quite a big change in behavior, no? Until now if users used a mux cache and a request matched no condition, it would just never get a response, which imo is a valid behavior. Now it gets an error?

File: pkg/cache/v3/simple_test.go
Comment (atollena) [inline pkg/cache/v3/simple_test.go]: There are a lot of `var streamState SubscriptionState` remaining. It would be good to clean them up. If you take my suggestion to just call this a `Subscription`, then this will make the code more terse (you can call the vars "sub") while retaining the clarity.

File: pkg/server/delta/v3/server.go
Comment (atollena) [inline pkg/server/delta/v3/server.go]: same remark regarding streamState -> subscriptionState (or streamState -> subscription or streamState -> sub), this time for non-test code.

File: pkg/server/sotw/v3/ads.go
Comment (atollena) [inline pkg/server/sotw/v3/ads.go]: nit: The project uses `TODO(name)` throughout.

```suggestion
			// TODO(valerian-roche): track ACK through subscription state
```

By the way, isn't this todo what you are doing here at L117? If not, I don't really understand this comment.

File: pkg/server/sotw/v3/server.go
Comment (atollena) [inline pkg/server/sotw/v3/server.go]: fwiw I found the code quite hard to read without the rename of this to something like `subscriptions`. There is Stream, StreamWrapper, StreamState which is actually a SubscriptionState...

File: pkg/server/sotw/v3/xds.go
Comment (atollena) [inline pkg/server/sotw/v3/xds.go]: Does that properly handle the case of unsubscribing from explicit wildcard? Probably this is never used, but still I think this violates the protocol spec and should be implemented properly. For example two subsequent requests:

```
Req(resources: ["*"]) -> explicit wildcard subscription -> IsWildcard() = true
...
Req(resources: []) -> IsWilcard = true, req.ResourceName = 0, GetSubscribedResources() == 0, 
```

This will fall into this if statement, even though it should not be considered wildcard and instead should be considered unsubscribing from wildcard.

There seem to be no way to implement the legacy wildcard behavior without storing an additional boolean in what you called `SubscriptionState`:

> For historical reasons, if the client sends a request for a given resource type but has never explicitly subscribed to any resource names

You can't know that without storing whether you've ever seen a resource name on the stream, can you? Without storing this explicitly, there is a high chance that you end up mixing up legacy/implicit and explicit wildcards.

File: pkg/server/stream/v3/subscription.go
Comment (atollena) [inline pkg/server/stream/v3/subscription.go]: Is this true? From what I can see this is also used in sotw via updateSubscriptionResources.

File: pkg/cache/v3/cache.go
Comment (valerian-roche) [reply pkg/cache/v3/cache.go]: Agreed, I think it makes sense

File: pkg/cache/v3/cache.go
Comment (valerian-roche) [reply pkg/cache/v3/cache.go]: Updated

File: pkg/cache/v3/cache.go
Comment (valerian-roche) [reply pkg/cache/v3/cache.go]: Updated

File: pkg/cache/v3/cache.go
Comment (valerian-roche) [reply pkg/cache/v3/cache.go:58]: Might be my English understanding, but calling "wildcard" makes me more believe it'd be returning something a bit random more than whether it's wildcard. Can update if preferred though

File: pkg/cache/v3/linear.go
Comment (valerian-roche) [reply pkg/cache/v3/linear.go:303]: We had issues on slack (and I think GitHub) due to such kind of silent errors.

File: pkg/cache/v3/mux.go
Comment (valerian-roche) [reply pkg/cache/v3/mux.go:43]: In the current case it gets a nil response, which will likely crash in the server implementation side though if I'm not mistaken

File: pkg/cache/v3/simple_test.go
Comment (valerian-roche) [reply pkg/cache/v3/simple_test.go]: Good remark, updated

File: pkg/server/delta/v3/server.go
Comment (valerian-roche) [reply pkg/server/delta/v3/server.go]: Updated

File: pkg/server/sotw/v3/ads.go
Comment (valerian-roche) [reply pkg/server/sotw/v3/ads.go]: Cleaned up, was leftover

File: pkg/server/sotw/v3/server.go
Comment (valerian-roche) [reply pkg/server/sotw/v3/server.go]: I am reworking this part. I believe the watch and the stream state ended up duplicating the same notion

File: pkg/server/stream/v3/subscription.go
Comment (valerian-roche) [reply pkg/server/stream/v3/subscription.go]: Nope will update as no longer true

File: pkg/server/sotw/v3/xds.go
Comment (valerian-roche) [reply pkg/server/sotw/v3/xds.go]: I agree that we do not support the case when the subscription was explicit wildcard and now requests nothing (not wildcard nor resources). The comment below mentions that we do not support this explicit case currently (in sotw or in delta), but I also don't see much value here. I can add this if preferred though

File: pkg/cache/v3/cache.go
Comment (atollena) [reply pkg/cache/v3/cache.go:58]: OK, this one is fine.

File: pkg/cache/v3/linear.go
Comment (atollena) [reply pkg/cache/v3/linear.go:303]: OK, I'll let you discuss this with other maintainers.

File: pkg/server/sotw/v3/server.go
Comment (atollena) [reply pkg/server/sotw/v3/server.go]: Yes, indeed. Closing this conversation and will re-review that part in https://github.com/DataDog/go-control-plane/pull/7.

**Review by atollena — CHANGES_REQUESTED** [review]

**Comment by atollena** [comment]
I guess you can close this PR.

File: pkg/cache/v3/mux.go
Comment (atollena) [reply pkg/cache/v3/mux.go:43]: Yes.

**Pull request closed without merging**
